### PR TITLE
Fix schema build for nested dataclasses / TypedDicts with discriminators

### PR DESCRIPTION
**Pull request by sydney-runkle**
Revisit the changes made in https://github.com/pydantic/pydantic/pull/7646, which introduced a few bugs for edge cases when building schemas for nested dataclasses / TypedDicts with discriminators.

Fix https://github.com/pydantic/pydantic/issues/8271
Fix https://github.com/pydantic/pydantic/issues/8628

Selected Reviewer: @hramezani

### CODE REVIEW

**Comment by sydney-runkle** [comment]
Please review

File: pydantic/_internal/_discriminated_union.py [hunk 1]
```diff
@@ -185,16 +187,27 @@ def apply(self, schema: core_schema.CoreSchema) -> core_schema.CoreSchema:
...
+
         assert not self._used
         schema = self._apply_to_root(schema)
         if self._should_be_nullable and not self._is_nullable:
             schema = core_schema.nullable_schema(schema)
         self._used = True
-        new_defs = collect_definitions(schema)
-        missing_defs = self.definitions.keys() - new_defs.keys()
+
+        # If there are any defintitions that were present on the original schema but not on the new schema,
```
Comment (dmontagu) [inline pydantic/_internal/_discriminated_union.py:201]:
```suggestion
        # If there are any definitions that were present on the original schema but not on the new schema,
```

File: pydantic/_internal/_discriminated_union.py [hunk 2]
```diff
@@ -185,16 +187,27 @@ def apply(self, schema: core_schema.CoreSchema) -> core_schema.CoreSchema:
...
+        # we need to add them to the new schema. This is necessary because the definitions may contain
+        # schemas that are referenced by the choices in the union, and we need to ensure that the new schema
+        # contains all the definitions that are necessary to resolve these references.
+        # Note -- by "original schema", we refer to the schema that was passed to the apply method,
+        # not the outermost schema that we're reucrsing on (where self.definitions came from).
```
Comment (dmontagu) [inline pydantic/_internal/_discriminated_union.py:206]:
```suggestion
        # not the outermost schema that we're recursing on (where self.definitions came from).
```

**Review by dmontagu — APPROVED** [review]

**Review by adriangb — APPROVED** [review]
If this fixes the new test cases and doesn't break others this seems good to me. Great job on the comments by the way!

Do we have any sort of benchmark _tests_ for this stuff? My main concern with future changes is that evidently modifying 1 line of code can take us from linear to exponential complexity. Maybe we can add a really naive test that just calls `time()` on `1` input and `time()` on `1000` input and make sure it's <= 5k times slower (so give or take linear)?

**Comment by sydney-runkle** [comment]
@adriangb,

Good question. To my knowledge, we don't have extensive performance testing for these things. I've opened an issue here: https://github.com/pydantic/pydantic/issues/8953. I'd like to address this issue before the patch release so that we can verify that we've improved things significantly.

**Pull request merged**